### ISSUE TITLE
Add a non-package-mode

### ISSUE THREAD

**Issue by johnthagen** [astral-sh/uv#10204]
### Proposal

For projects that will not be published as wheels but simply want to track/lock their dependencies with `uv`, it would be nice to be able to omit the following lines from `pyproject.toml`:

```toml
[project]
name = "my-project"
version = "0.1.0"
description = "Add your description here"
readme = "README.md"
```

Poetry has this feature:

- https://python-poetry.org/docs/basic-usage/#operating-modes

```toml
[tool.poetry]
package-mode = false
```

### Configuration Example

A straw-man proposal for how this _could_ be configured in uv:

```toml
[project]
requires-python = ">=3.12"
dependencies = [
    "mkdocs>=1.6.1",
]

[tool.uv]
package-mode = false
```

### Motivation

A long thread about motivation for this is in the original Poetry issue

- https://github.com/python-poetry/poetry/issues/1132

In a nutshell, there are many types of Projects where specifying an extra version, name, description, etc is unnecessary and extra boilerplate. For example:

- A project that uses MkDocs to build a static site and has no Python code in it all (but the user still needs to lock the version of MkDocs and its dependencies for reproducibility)
- A Django or FastAPI application where the execution mode is usually something like `uvicorn main:app` or `python runserver` where no package/wheel is installed
- A collection of standalone Python files that have requirements but aren't organized into a package installed into a venv

An example referencing the boilerplate in the wild:

- https://gitlab.jhuapl.edu/PCBDR/pcbdr/-/tree/main/docs/user_guide#build

> Our Django projects won’t really care about the `name`, `version`, `description`, and `readme` properties, so just leave them as is.

### Challenges

The complication I see with `uv` as compared to Poetry is that perhaps it's not technically _legal_ to omit certain fields from the `[project]` field? But if a project isn't actually defining a true package/wheel perhaps that doesn't matter?

Currently if you run `uv sync` without these `[project]` fields, you get an error:

```
error: Failed to parse: `pyproject.toml`
  Caused by: TOML parse error at line 1, column 1
  |
1 | [project]
  | ^^^^^^^^^
`pyproject.toml` is using the `[project]` table, but the required `project.name` field is not set
```

**Comment by johnthagen** [comment]
@T-256 Using your `pyproject.toml` without a `[project]` field, `uv sync` fails:

```
$ uv sync
error: No `project` table found in: `/pyproject.toml`
```

**Comment by charliermarsh** [comment]
If you add `[tool.uv.workspace]` to it, it won't fail.

**Comment by johnthagen** [comment]
Thanks @charliermarsh! Is `[tool.uv.workspace]` intended to solve the use case described in this issue?

Reading the workspace docs, it seems like more of a multi-project feature rather than a "single project but I don't need the wheel project metadata" feature?

- https://docs.astral.sh/uv/concepts/projects/workspaces/#using-workspaces

I ask just as I think through how I would explain this to someone else using single Django or FastAPI application.

Also, when I run this way, even with a `.python-version` I get a warning:

```
warning: No `requires-python` value found in the workspace. Defaulting to `>=3.13`.
```

And it seems that `requires-python` is not allowed at the `[tool.uv.workspace]` level.

**Comment by johnthagen** [comment]
> I also have some environment usages where only need declared dependencies. Rye's [virtual project](https://rye.astral.sh/guide/virtual/#workspaces):

@T-256 Oh nice. I haven't used Rye before, but that looks like it's filling the same use case as Poetry's `package-mode = false`.

**Comment by charliermarsh** [comment]
Yeah we might improve it over time, but right now it _is_ recommended that you create a `[project]` table. The example you gave would actually violate standards:

```toml
[project]
requires-python = ">=3.12"
dependencies = [
    "mkdocs>=1.6.1",
]
```

The standard declares that you _must_ have at least a package name and a version (or a dynamic version) -- this isn't a uv thing, it's a standards thing.

**Comment by johnthagen** [comment]
I was curious how Rye handled this. When passing `--virtual` it still fills out of the `[project]` fields. But Rye doesn't require that you keep them (it doesn't seem to validate the `[project]` field, not sure if that is intentional or not). This `pyproject.toml` works with Rye 0.43.0:

```toml
[project]
dependencies = [
    "mkdocs<1.5",
]
requires-python = ">= 3.8"

[tool.rye]
managed = true
virtual = true
dev-dependencies = []
```

```
$ rye sync
Reusing already existing virtualenv
Generating production lockfile: /home/user/Downloads/requirements.lock
Generating dev lockfile: /home/user/Downloads/requirements-dev.lock
Installing dependencies
Resolved 13 packages in 4ms
Audited 13 packages in 0.02ms
Done!
```

Poetry gets around the issue (currently) because it hasn't released a version that supports `[project]` table, thus this works:

```toml
[tool.poetry]
package-mode = false

[tool.poetry.dependencies]
python = "^3.11"

mkdocs = "*"
```

I haven't tested Poetry 2.0 which supports `[project]` in their main branch but hasn't been released. They may run into similar problems for users wanting to use the standard `[project]` table and omitting unneeded boilerplate if they'd want to enforce standards-compliance.

In any case, I definitely appreciate the challenge of making this work smoothly given the current standards.

**Comment by edmorley** [comment]
I feel like the [pyproject.toml spec](https://packaging.python.org/en/latest/specifications/pyproject-toml/#declaring-project-metadata-the-project-table) is more restrictive than it should be (given it states `name` and `version` are mandatory). (Particularly since the spec purports to be based on the decisions in [PEP 518](https://peps.python.org/pep-0518/) and [PEP 621](https://peps.python.org/pep-0621/), but neither of those peps explicitly said `name` and `version` should be mandatory.)

I wonder if upstream would be open to revising the spec to state that `name` and `version` are only required when building the project using a build backend? (Either the explicitly specified `[build-system]` or else the implicit one.) 

Adjusting the spec in that way would actually be in keeping with other parts of the spec, which already state things like:

> Tools should not require the existence of the [build-system] table. A pyproject.toml file may be used to store configuration details other than build-related data and thus lack a [build-system] table legitimately.

(in https://packaging.python.org/en/latest/specifications/pyproject-toml/#declaring-build-system-dependencies-the-build-system-table)

**Comment by johnthagen** [comment]
Related existing feature

- https://docs.astral.sh/uv/concepts/projects/config/#build-systems

> uv uses the presence of a build system to determine if a project contains a package that should be installed in the project virtual environment. If a build system is not defined, uv will not attempt to build or install the project itself, just its dependencies. If a build system is defined, uv will build and install the project into the project environment.

**Comment by T-256** [comment]
With https://github.com/astral-sh/uv/pull/13735 we can now specify required python without using `project` table (package mode):
```toml
[tool.uv.dependency-groups.dev]
requires-python = ">=3.12"

[dependency-groups]
dev = [
    "mkdocs>=1.6.1",
]
```

UPD: with #13742 we can also avoid writing`[tool.uv.workspace]`.

**Comment by mikeweltevrede** [comment]
Have you also considered the `package=false` setting? https://docs.astral.sh/uv/reference/settings/#package

```toml
[tool.uv]
package = false
```

I am not sure if it fully covers your needs though.

**Comment by zanieb** [comment]
Is there more to do here still?

**Comment by johnthagen** [comment]
@zanieb I think this can be closed. Ideally, I'd like to be able to omit many of the `[project]` fields for non-wheel projects where they aren't needed, like a FastAPI app (and can become needless boilerplate), but that would not be compliant with PEP 621. Rye [seems to allow this anyway](https://github.com/astral-sh/uv/issues/10204#issuecomment-2563996532) but I'm not sure if `uv` wants to encourage this behavoir?

I'd say this issue can probably be closed as their are good reasons for the status quo and `package = false` solves the most important needs.

**Comment by vmgustavo** [comment]
Is there a flag for `uv init` that initializes a `pyproject.toml` in a non package mode? If no, would it be possible to add?

Edit: change non project to non package

**Comment by mikeweltevrede** [comment]
@vmgustavo what defines a "non-project" mode, i.e. what would make it different and what would you need from it? Aren't all repos a project or sorts? That's also why the name "pyproject.toml" was chosen in PEP-621.

**Comment by vmgustavo** [comment]
Sorry @mikeweltevrede it was supposed to be non package mode.

**Comment by mikeweltevrede** [comment]
@vmgustavo Okay, clear! Then that can be achieved with the `package` setting: https://docs.astral.sh/uv/reference/settings/#package

**Comment by vmgustavo** [comment]
Yes and if I create a pyproject file using `uv init` I get the default package oriented file. If I simply remove the contents of the file and add the package flag the error `No "project" table found` appears. I assume there should be more sections that must be added to make this work and if it could be created with the `uv sync` command with a flag that would be awesome.

I see there is already a `--no-package` flag

```
--no-package   Do not set up the project to be built as a Python package
```

However the resulting pyproject is the same as when I generate it without this flag. I don't know if I am doing something wrong but this is what I get:

```
> uv init
Initialized project `uvt`
> cat pyproject.toml
[project]
name = "uvt"
version = "0.1.0"
description = "Add your description here"
readme = "README.md"
requires-python = ">=3.12"
dependencies = []
```

```
> uv init --no-project
Initialized project `uvt2`
> cat pyproject.toml
[project]
name = "uvt2"
version = "0.1.0"
description = "Add your description here"
readme = "README.md"
requires-python = ">=3.12"
dependencies = []
```

**Comment by mikeweltevrede** [comment]
@vmgustavo Notice in your second command you do `uv init --no-project` and not `uv init --no-package`.

@zanieb (tagging you since you asked if there is more to be done) By the way this seems to have a bug: `--no-project` is not listed as a supported flag for `uv init` yet it does not error continues indeed as if the user ran a bare `uv init`:

```shell
C:\...\uvex>uv init --no-project
Initialized project `uvex`
```

**Comment by vmgustavo** [comment]
Sorry again for the mistake, but the same happens for the `--no-package` flag.

```
> uv init --no-package
Initialized project `uvt3`
> cat pyproject.toml
[project]
name = "uvt3"
version = "0.1.0"
description = "Add your description here"
readme = "README.md"
requires-python = ">=3.12"
dependencies = []
```

**Comment by zanieb** [comment]
`--no-project` is about disabling project discovery, it's a global CLI flag, it doesn't have anything to do with the `[project]` table.

And, yeah, `uv init --no-package` still writes a `[project]` table for dependencies. That's intentional. We could consider not... and have `uv add` write somewhere else? but that's a big change.

I don't think we'll support spec-incompliant `[project]` tables with missing fields.

**Comment by cdwilson** [comment]
I just found this while searching for how to manage python dependencies with uv for a MkDocs documentation repo that is not a python package.

Piecing things together from the comments above, it appears that the current solution is to use a pyproject.toml like the following:

```toml
[tool.uv]
package = false

[tool.uv.dependency-groups.docs]
requires-python = ">=3.13"

[dependency-groups]
docs = [
    "mkdocs-material>=9.6.21",
]
```

I thought from https://github.com/astral-sh/uv/issues/10204#issuecomment-2978496593 that the project table might not be needed, however, this results in the following warning:

```plaintext
❯ uv add --group docs mkdocs-material
warning: No `requires-python` value found in the workspace. Defaulting to `>=3.13`.
Resolved 29 packages in 142ms
Bytecode compiled 810 files in 523ms
```

Adding a `project` table causes the warning to go away. 

```toml
[project]
name = "docs"
version = "0.1.0"
requires-python = ">=3.13"

[tool.uv]
package = false

[tool.uv.dependency-groups.docs]
requires-python = ">=3.13"

[dependency-groups]
docs = [
    "mkdocs-material>=9.6.21",
]
```

For anyone else who is doing something similar, is this still the recommended way to manage dependencies for projects that aren't python packages?

**Comment by konstin** [comment]
@cdwilson Tracking this in a separate issue: https://github.com/astral-sh/uv/issues/16136

**Comment by amluto** [comment]
I'm looking at this issue and #10208 (which links here), and it looks like `package = false` doesn't quite do it.  I have an example here (not on PyPI yet, but I intend to publish it soon):

https://github.com/amluto/lutron-integration/blob/05dc5a8319de51d5443ea958e5ba2b3d820fcbd4/pyproject.toml

I'll quote it here for reference:

```
[project]
name = "lutron-integration-workspace-root"
version = "0.0.0"
requires-python = ">=3.12"
dependencies = [
    "lutron-integration",
    "lutron-integration-tools",
]

[tool.uv]
package = false

[tool.uv.workspace]
members = ["packages/*"]

[tool.uv.sources]
lutron-integration-tools = { workspace = true }
lutron-integration = { workspace = true }

[dependency-groups]
dev = [
    "ruff>=0.14.4",
    "pytest>=8.4.2",
    "ty>=0.0.5",
]
```

It's doing kind of well.  `uv sync` and `uv run` do the right thing.  Even ty seems to do the right thing, although the docs make it sounds like I can't really trust it to work right until ty is a bit farther along.  It's silly that I needed to name the project and give it a version, but I can live with it.  Putting dependencies here makes sense.

But #10208 points out that the project isn't suppose to build.  But `uv build` builds a source package and a wheel, which is ridiculous.  (I'm a bit baffled here -- what build backend is getting used?  I didn't specify one.)

If I add:

```
[build-system]
requires = ["uv_build >= 0.9.8, <0.10.0"]
build-backend = "uv_build"
```

to pyproject.toml, then `uv run` and `uv sync` still work, but `uv build` fails like so:

```
% uv build
Building source distribution...
Error: Expected a Python module at: src/lutron_integration_workspace_root/__init__.py
  × Failed to build `/Volumes/devel/hass/pylutron_integration`
  ├─▶ The build backend returned an error
  ╰─▶ Call to `uv_build.build_sdist` failed (exit status: 1)
      hint: This usually indicates a problem with the package or the build environment.
```

On the one hand, great -- building this makes no sense.  On the other hand, this is absurd.

It seems to me that a farily straightforward way to make this robust would be to teach uv_build to notice the `package = false` setting and politely refuse to build the package, and to document that the recommended way to use package = false is to pair it with setting the build system to uv-build.  And then `uv build` would fail with a good error message and other PEP 517 frontends would also refuse to build the package even though they don't understand `package = false`.

There's another amusing solution:

```
[build-system]
requires = ["uv_build >= 0.9.8, <0.8.0"]
build-backend = "uv_build"
```

Yes, I intentionally corrupted the `requires` so that it can't possibly be satisfied.  Now `uv build` fails very quickly indeed :)  And I suppose anyone could go publish a fully PEP 517 compliant build backend that simply refuses to build anything, but ISTM uv_build is the perfect place for this logic.

**Comment by konstin** [comment]
Why does your top level `pyproject.toml` have a `[project]` table, does it work if it doesn't have one?

**Comment by amotl** [comment]
> The standard declares that you _must_ have at least a package name and a version (or a dynamic version) -- this isn't a uv thing, it's a standards thing.

Hi. I would like to present a total different use case which I still think should be considered in this discussion: We are maintaining projects that still use `setup.py` files to store primary project metadata, while accompanying `pyproject.toml` files store auxiliary project metadata.

We've reported about it over here, because we also would like to get rid of the runtime warning:
> No `requires-python` value found in the workspace.

- https://github.com/astral-sh/uv/issues/16136#issuecomment-3969844401
